### PR TITLE
support stark aggregation

### PR DESCRIPTION
**Pull request by cyl19970726**
first run
```
./stark_aggregation.sh yes
``` 

second run 
```
./stark_aggregation.sh 
```

### CODE REVIEW

File: starkjs/src/recursive/main_joinzkin.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
+const fs= require("fs");
+const path = require("path");
+const { polMulAxi } = require("pil-stark/src/polutils");
+const version = require("../../package").version;
+const JSONbig = require("json-bigint");
```
Comment (eigmax) [inline starkjs/src/recursive/main_joinzkin.js:5]: is this dead code?

File: starky/src/stark_verifier_circom.rs [hunk 2]
```diff
@@ -1589,24 +1589,292 @@ template StarkVerifier() {{
...
     ));
 
+    ///////
+    // Aggregation Stage
+    ///////
+
+    if options.agg_stage {
+        res.push_str(&format!(
+            r#"
+template Recursive2() {{
```
Comment (eigmax) [inline starky/src/stark_verifier_circom.rs:1607]: This is actually recursivef?

Comment (cyl19970726) [reply starky/src/stark_verifier_circom.rs:1607]: It is recursive2 because it combines two starkVerifier() into one circuit.

Comment (cyl19970726) [reply starky/src/stark_verifier_circom.rs:1607]: [here ](https://github.com/0xEigenLabs/eigen-zkvm/pull/71/files#diff-07e89d24351a3dc427f8f460d0e901d0426d393c388b4e4fc7d5d569982a4dd8R82) generates recursivef

File: test/stark_aggregation.sh [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    -i $input0/r1_input.zkin.json  \
+    -p $WORKSPACE/$RECURSIVE_CIRCUIT.pil  \
+    -e $WORKSPACE/$RECURSIVE_CIRCUIT.exec \
+    -m $WORKSPACE/$RECURSIVE_CIRCUIT.cm
+
+echo "5. generate recursive2 proof  "
+# generate the stark proof and the circom circuits to verify stark proof.
+# input files : $C12_VERIFIER.pil.json(stark proof)  $C12_VERIFIER.const(const polynomials)  $C12_VERIFIER.cm (commit polynomials)
+# output files :  $RECURSIVE2_CIRCUIT.circom  $RECURSIVE2_CIRCUIT/r2_input.json
+# Remark: the N of r2.starkStruct must be 2^20 , because the degree of $RECURSIVE_CIRCUIT.pil is 2^20 which determined by the proocess of converting  $RECURSIVE_CIRCUIT.circom to  $RECURSIVE_CIRCUIT.pil
```
Comment (eigmax) [inline test/stark_aggregation.sh:71]: should do `mkdir -p ./aggregation/$RECURSIVE2_CIRCUIT`

File: test/stark_aggregation.sh [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    --o $WORKSPACE/$RECURSIVE_CIRCUIT.const \
+    --m $WORKSPACE/$RECURSIVE_CIRCUIT.cm -c $RUNDIR/circuits/$RECURSIVE2_CIRCUIT.circom --i ./aggregation/$RECURSIVE2_CIRCUIT/r2_input.zkin.json 
+
+
+
+# final recursive stage 
+echo " ==> final recursive stage <== "
+echo "1. compile circuit and generate r1cs and wasm"
+${ZKIT} compile -p goldilocks -i $RUNDIR/circuits/$RECURSIVE2_CIRCUIT.circom -l "../starkjs/node_modules/pil-stark/circuits.gl" -l "../starkjs/node_modules/circomlib/circuits" --O2=full -o $WORKSPACE 
+
```
Comment (eigmax) [inline test/stark_aggregation.sh:83]: OOM raised, set:  export NODE_OPTIONS="--max-old-space-size=51200"

File: test/stark_aggregation.sh [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+echo "2. generate the pil files and  const polynomicals files "
+node $RUNDIR/src/compressor12/main_compressor12_setup.js \
```
Comment (eigmax) [inline test/stark_aggregation.sh:85]: Out of memory, should export NODE_OPTIONS="--max-old-space-size=10240".

File: test/recursive_proof_to_stark.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    -p $WORKSPACE/$C12_VERIFIER.pil  \
+    -e $WORKSPACE/$C12_VERIFIER.exec \
+    -m $WORKSPACE/$C12_VERIFIER.cm
+
+mkdir -p ./aggregation/$RECURSIVE1_VERIFIER/
+
+# generate the stark proof and the circom circuits to verify stark proof.
+# input files : $C12_VERIFIER.pil.json(stark proof)  $C12_VERIFIER.const(const polynomials)  $C12_VERIFIER.cm (commit polynomials)
+# output files :  $RECURSIVE1_VERIFIER.circom  $RECURSIVE1_VERIFIER/input.json
+../target/release/eigen-zkit stark_prove -s ../starky/data/c12.starkStruct.json \
```
Comment (eigmax) [inline test/recursive_proof_to_stark.sh:50]: Merge this file to recursive_proof_to_snark?

File: test/stark_aggregation.sh [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+NUM_PROOF=2
+NUM_INPUT=2
+CUR_DIR=$(cd $(dirname $0);pwd)
+ZKIT="${CUR_DIR}/../target/release/eigen-zkit"
+CIRCUIT="fibonacci"
+PILEXECJS="fibonacci/fibonacci.js"
+RUNDIR="${CUR_DIR}/../starkjs"
+
+WORKSPACE=/tmp/aggregation_$CIRCUIT
+if [ $1 = "restart" ]; then
```
Comment (eigmax) [inline test/stark_aggregation.sh:20]: force=${1-no}
if [ $force = "yes" ]; then
    rm -rf $WORKSPACE && mkdir -p $WORKSPACE
fi

File: starkjs/package.json [hunk 8]
```diff
@@ -16,6 +16,7 @@
     "circomlib": "^2.0.5",
     "circomlibjs": "^0.1.6",
     "ffjavascript": "^0.2.59",
+    "fs": "^0.0.1-security",
```
Comment (eigmax) [inline starkjs/package.json:19]: Why do we need  this deps?

File: starkjs/src/recursive/main_joinzkin.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    zkinOut.b_s2_siblings = zkin2.s2_siblings;
+    zkinOut.b_s3_siblings = zkin2.s3_siblings;
+    zkinOut.b_s4_siblings = zkin2.s4_siblings;
+    zkinOut.b_s1_vals = zkin2.s1_vals;
+    zkinOut.b_s2_vals = zkin2.s2_vals;
+    zkinOut.b_s3_vals = zkin2.s3_vals;
+    zkinOut.b_s4_vals = zkin2.s4_vals;
+    zkinOut.b_finalPol = zkin2.finalPol;
+    
+    await fs.promises.writeFile(zkinOutFile, JSON.stringify(zkinOut, null, 1), "utf8");
```
Comment (eigmax) [inline starkjs/src/recursive/main_joinzkin.js:78]: use standard fs.  fs.writeFileSync

File: test/snark_verifier.sh [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        snarkjs zkev  $WORK_DIR/g16.zkey  $WORK_DIR/verification_key.json
+
+        echo "4. verify groth16 proof"
+        snarkjs g16v $WORK_DIR/verification_key.json $WORK_DIR/public.json $WORK_DIR/proof.json
+
+        cp $WORK_DIR/public.json /tmp/aggregation/final_public.json 
+        cp $WORK_DIR/proof.json /tmp/aggregation/final_proof.json
+
+        echo "5. generate verifier contract"
+        snarkjs zkesv  $WORK_DIR/g16.zkey  ${CUR_DIR}/aggregation/contracts/final_verifier.sol
```
Comment (eigmax) [inline test/snark_verifier.sh:64]: we can export snarkjs=$CUR/aggregation/.bin/snarkjs after run `npm i` under $CUR/aggregation

**Comment by cyl19970726** [comment]
C12 Stage Time Cost (38s)
Aggregation Stage Time Cost (51s)
Final Stage Time Cost (30s)
Recursive Snark Stage Time Cost (47s)
Full Process Time Cost (166s)

**Comment by cyl19970726** [comment]
on chain verify gas  cost: 40w gas

**Comment by cyl19970726** [comment]
AMD Ryzen 9 5950X 16-Core Processor
16G memory

**Comment by eigmax** [comment]
this is for g16? @cyl19970726

**Pull request merged**